File: pylox/pylox/Expr.py
from __future__ import annotations
from abc import ABC, abstractmethod
from typing import Final, override

from Token import Token


class Expr(ABC):
    @abstractmethod
    def accept[R](self, visitor: Visitor[R]) -> R: ...


class Visitor[R]:
    @abstractmethod
    def visit_Assign_Expr(self, expr: Assign) -> R: ...

    @abstractmethod
    def visit_Binary_Expr(self, expr: Binary) -> R: ...

    @abstractmethod
    def visit_Call_Expr(self, expr: Call) -> R: ...

    @abstractmethod
    def visit_Get_Expr(self, expr: Get) -> R: ...

    @abstractmethod
    def visit_Grouping_Expr(self, expr: Grouping) -> R: ...

    @abstractmethod
    def visit_Literal_Expr(self, expr: Literal) -> R: ...

    @abstractmethod
    def visit_Logical_Expr(self, expr: Logical) -> R: ...

    @abstractmethod
    def visit_Set_Expr(self, expr: Set) -> R: ...

    @abstractmethod
    def visit_Unary_Expr(self, expr: Unary) -> R: ...

    @abstractmethod
    def visit_This_Expr(self, expr: This) -> R: ...

    @abstractmethod
    def visit_Variable_Expr(self, expr: Variable) -> R: ...


class Assign(Expr):
    def __init__(self, name: Token, value: Expr):
        super().__init__()
        self.name: Final[Token] = name
        self.value: Final[Expr] = value

    @override
    def accept[R](self, visitor: Visitor[R]) -> R:
        return visitor.visit_Assign_Expr(self)


class Binary(Expr):
    def __init__(self, left: Expr, operator: Token, right: Expr):
        super().__init__()
        self.left: Final[Expr] = left
        self.operator: Final[Token] = operator
        self.right: Final[Expr] = right

    @override
    def accept[R](self, visitor: Visitor[R]) -> R:
        return visitor.visit_Binary_Expr(self)


class Call(Expr):
    def __init__(self, callee: Expr, paren: Token, arguments: list[Expr]):
        super().__init__()
        self.callee: Final[Expr] = callee
        self.paren: Final[Token] = paren
        self.arguments: Final[list[Expr]] = arguments

    @override
    def accept[R](self, visitor: Visitor[R]) -> R:
        return visitor.visit_Call_Expr(self)


class Get(Expr):
    def __init__(self, object: Expr, name: Token):
        super().__init__()
        self.object: Final[Expr] = object
        self.name: Final[Token] = name

    @override
    def accept[R](self, visitor: Visitor[R]) -> R:
        return visitor.visit_Get_Expr(self)


class Grouping(Expr):
    def __init__(self, expression: Expr):
        super().__init__()
        self.expression: Final[Expr] = expression

    @override
    def accept[R](self, visitor: Visitor[R]) -> R:
        return visitor.visit_Grouping_Expr(self)


class Literal(Expr):
    def __init__(self, value: object):
        super().__init__()
        self.value: Final[object] = value

    @override
    def accept[R](self, visitor: Visitor[R]) -> R:
        return visitor.visit_Literal_Expr(self)


class Logical(Expr):
    def __init__(self, left: Expr, operator: Token, right: Expr):
        super().__init__()
        self.left: Final[Expr] = left
        self.operator: Final[Token] = operator
        self.right: Final[Expr] = right

    @override
    def accept[R](self, visitor: Visitor[R]) -> R:
        return visitor.visit_Logical_Expr(self)


class Set(Expr):
    def __init__(self, object: Expr, name: Token, value: Expr):
        super().__init__()
        self.object: Final[Expr] = object
        self.name: Final[Token] = name
        self.value: Final[Expr] = value

    @override
    def accept[R](self, visitor: Visitor[R]) -> R:
        return visitor.visit_Set_Expr(self)


class Unary(Expr):
    def __init__(self, operator: Token, right: Expr):
        super().__init__()
        self.operator: Final[Token] = operator
        self.right: Final[Expr] = right

    @override
    def accept[R](self, visitor: Visitor[R]) -> R:
        return visitor.visit_Unary_Expr(self)


class This(Expr):
    def __init__(self, keyword: Token):
        super().__init__()
        self.keyword: Final[Token] = keyword

    @override
    def accept[R](self, visitor: Visitor[R]) -> R:
        return visitor.visit_This_Expr(self)


class Variable(Expr):
    def __init__(self, name: Token):
        super().__init__()
        self.name: Final[Token] = name

    @override
    def accept[R](self, visitor: Visitor[R]) -> R:
        return visitor.visit_Variable_Expr(self)

File: golox/internal/expr/expr.go
package expr

import "github.com/englishlayup/crafting_interpreters/golox/internal/token"

type Expr[R any] interface {
	Accept(visitor ExprVisitor[R]) R
}

type ExprVisitor[R any] interface {
	VisitAssignExpr(expr Assign[R]) R
	VisitBinaryExpr(expr Binary[R]) R
	VisitCallExpr(expr Call[R]) R
	VisitGetExpr(expr Get[R]) R
	VisitGroupingExpr(expr Grouping[R]) R
	VisitLiteralExpr(expr Literal[R]) R
	VisitLogicalExpr(expr Logical[R]) R
	VisitSetExpr(expr Set[R]) R
	VisitSuperExpr(expr Super[R]) R
	VisitUnaryExpr(expr Unary[R]) R
	VisitThisExpr(expr This[R]) R
	VisitVariableExpr(expr Variable[R]) R
}

type Assign[R any] struct {
	Name  token.Token
	Value Expr[R]
}

func (a Assign[R]) Accept(visitor ExprVisitor[R]) R {
	return visitor.VisitAssignExpr(a)
}

type Binary[R any] struct {
	Left     Expr[R]
	Operator token.Token
	Right    Expr[R]
}

func (b Binary[R]) Accept(visitor ExprVisitor[R]) R {
	return visitor.VisitBinaryExpr(b)
}

type Call[R any] struct {
	Callee    Expr[R]
	Paren     token.Token
	Arguments []Expr[R]
}

func (c Call[R]) Accept(visitor ExprVisitor[R]) R {
	return visitor.VisitCallExpr(c)
}

type Get[R any] struct {
	Object Expr[R]
	Name   token.Token
}

func (g Get[R]) Accept(visitor ExprVisitor[R]) R {
	return visitor.VisitGetExpr(g)
}

type Grouping[R any] struct {
	Expression Expr[R]
}

func (g Grouping[R]) Accept(visitor ExprVisitor[R]) R {
	return visitor.VisitGroupingExpr(g)
}

type Literal[R any] struct {
	Value any
}

func (l Literal[R]) Accept(visitor ExprVisitor[R]) R {
	return visitor.VisitLiteralExpr(l)
}

type Logical[R any] struct {
	Left     Expr[R]
	Operator token.Token
	Right    Expr[R]
}

func (l Logical[R]) Accept(visitor ExprVisitor[R]) R {
	return visitor.VisitLogicalExpr(l)
}

type Set[R any] struct {
	Object Expr[R]
	Name   token.Token
	Value  Expr[R]
}

func (s Set[R]) Accept(visitor ExprVisitor[R]) R {
	return visitor.VisitSetExpr(s)
}

type Super[R any] struct {
	Keyword token.Token
	Method  token.Token
}

func (s Super[R]) Accept(visitor ExprVisitor[R]) R {
	return visitor.VisitSuperExpr(s)
}

type Unary[R any] struct {
	Operator token.Token
	Right    Expr[R]
}

func (u Unary[R]) Accept(visitor ExprVisitor[R]) R {
	return visitor.VisitUnaryExpr(u)
}

type This[R any] struct {
	Keyword token.Token
}

func (t This[R]) Accept(visitor ExprVisitor[R]) R {
	return visitor.VisitThisExpr(t)
}

type Variable[R any] struct {
	Name token.Token
}

func (v Variable[R]) Accept(visitor ExprVisitor[R]) R {
	return visitor.VisitVariableExpr(v)
}
